File: lua/HomingMissile.lua
-- Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt

if not HomingMissile then

HomingMissile							=	{ }
HomingMissile.className					=	"HomingMissile" 

HomingMissile.STATE_HMISSILE_TRACKING	=	1
HomingMissile.STATE_HMISSILE_FALLING	=	2
HomingMissile.STATE_HMISSILE_SLEEPING	=	3


HomingMissile.kHMISSILES_MAX					= 	8
HomingMissile.kHMISSILE_MAXTURNANGLE			=	90.0
HomingMissile.kHMISSILE_SPEED					=	200.0
HomingMissile.kHMISSILE_LIFESPAN				=	15.0
HomingMissile.kHMISSILE_ANGULAR_ACCELERATION	=	180.0


local	ANIM_ENEMYMISSILE_FRAME1	=	"enemy_missile__1"
local	ANIM_ENEMYMISSILE_FRAME2	=	"enemy_missile__2"
local	ANIM_ENEMYMISSILE_FRAME3	=	"enemy_missile__3"

local	ANIM_PLAYERMISSILE_FRAME1	=	"player_missile__1"
local	ANIM_PLAYERMISSILE_FRAME2	=	"player_missile__2"
local	ANIM_PLAYERMISSILE_FRAME3	=	"player_missile__3"


	function HomingMissile.Create( startpos,  rot,  follow,  damage, lifespan)

		--Logger.lprint("HomingMissile.Create()") 

		local homingmissile = {
            	following		=	follow,
				vPos			=	startpos and Vector4.Create(startpos) or 
									Vector4.Create(follow:GetVectorPosition():X()-50+math.random(0,100),follow:GetVectorPosition():Y()-720),
				vDir			=	Vector4.Create(0.7,0.7),
	            rot				=	rot or 0.0,
	            Lifespan		=	lifespan or HomingMissile.kHMISSILE_LIFESPAN,
	            time			=	0.0,
				strength		=	255,
	            nState			=	HomingMissile.STATE_HMISSILE_TRACKING,
	            damage			=	damage or 0,
	            bInvincible		=	false,
				bMarkedForDeath	=	false,
				isPlayer 		=	follow.className ~= 'Player', 
				colline 		=	Line2D.Create(Vector4.Create(0,0),Vector4.Create(0,0)),
	            currentAngularDir,
	            currentAngularSpeed,
	            tMissile,
				className		=	HomingMissile.className,

		}
		setmetatable(homingmissile, {
	  		__index = HomingMissile
		})

		homingmissile.vPos:Add(homingmissile.vPos,Vector4.Create(math.random(1,4),math.random(1,4)))
		homingmissile.animation = TextureAnim.Create(3)

		
		if (homingmissile.isPlayer) then
			homingmissile.animation:AddFrame(ANIM_PLAYERMISSILE_FRAME1)
			homingmissile.animation:AddFrame(ANIM_PLAYERMISSILE_FRAME2)
			homingmissile.animation:AddFrame(ANIM_PLAYERMISSILE_FRAME3)
		else
			homingmissile.animation:AddFrame(ANIM_ENEMYMISSILE_FRAME1)
			homingmissile.animation:AddFrame(ANIM_ENEMYMISSILE_FRAME2)
			homingmissile.animation:AddFrame(ANIM_ENEMYMISSILE_FRAME3)
		end

		homingmissile.animation:Play(true, true)

		return homingmissile 
	end

	function HomingMissile.Follow(homingmissile,who)
		homingmissile.following = who
	end
	
	function HomingMissile.Update(homingmissile,dt)
 		HomingMissile.UpdateLive(homingmissile,dt)
	end
	
	function HomingMissile.UpdateLive(homingmissile,dt)

		--Logger.lprint("HomingMissile:Update()") 

		assert(dt ~= 0,'HomingMissile - invalid Update dt')

		if (homingmissile.animation) then
		 	homingmissile.animation:Update(dt)
		end
		
		homingmissile.time =	homingmissile.time +	dt
		
		
		local	diff_vec	=	Temp.v0 
		local	state		=	homingmissile.nState
				
		if(state	~=	HomingMissile.STATE_HMISSILE_FALLING and homingmissile.time > homingmissile.Lifespan) then
		
			homingmissile.nState	=	HomingMissile.STATE_HMISSILE_FALLING
			state					=	homingmissile.nState
		
		end
			   		
		if(state	==	HomingMissile.STATE_HMISSILE_FALLING or not homingmissile.following:IsAlive()) then

			diff_vec:SetXyzw(0,homingmissile.vPos:Y(),0,0)
			if (homingmissile.time > 1.25*homingmissile.Lifespan) then
				homingmissile.nState	=	HomingMissile.STATE_HMISSILE_SLEEPING
				HomingMissile.MarkForDeath(homingmissile)
				homingmissile.time		=	0
				return
			end
		elseif(state	==	HomingMissile.STATE_HMISSILE_TRACKING) then

			Vector4.Subtract(diff_vec,homingmissile.following:GetVectorPosition(), homingmissile.vPos)

		elseif(state	==	HomingMissile.STATE_HMISSILE_SLEEPING) then
			if (homingmissile.time > 4) then
				homingmissile.nState	=	HomingMissile.STATE_HMISSILE_TRACKING
			end
			return
		end

		diff_vec:Normalise2()

		-- Work out current angle
		local rot_tar = math.atan2(diff_vec:Y(), diff_vec:X())
		rot_tar = math.deg(rot_tar)
		rot_tar = ( rot_tar + 360.0 ) % 360.0


		local heading_vec= Temp.v1  	
		
		heading_vec:SetXyzw(homingmissile.vDir:X(),homingmissile.vDir:Y(),0,0)
		heading_vec:Normalise2()

		local perpheading_vec= Temp.v2 

		perpheading_vec:SetXyzw(-heading_vec:Y(), heading_vec:X(), 0, 0)


		local perpCosineAngle = Vector4.Dot2(diff_vec, perpheading_vec)
		
		local newAngularDirection	=	(perpCosineAngle>0) and 1.0 or -1.0
		
		
		if ( newAngularDirection ~= homingmissile.currentAngularDir ) then
			homingmissile.currentAngularDir = newAngularDirection
			
			homingmissile.currentAngularSpeed = 0.0
		end

		homingmissile.currentAngularSpeed = homingmissile.currentAngularSpeed +(dt * HomingMissile.kHMISSILE_ANGULAR_ACCELERATION)

		if ( homingmissile.currentAngularSpeed > HomingMissile.kHMISSILE_MAXTURNANGLE ) then
			homingmissile.currentAngularSpeed = HomingMissile.kHMISSILE_MAXTURNANGLE
		end
		
		local maxTurnAngle = homingmissile.currentAngularSpeed * dt

		if(math.abs(rot_tar-homingmissile.rot)>maxTurnAngle) then
			local leftAngle  = ( homingmissile.rot + 360.0 - maxTurnAngle ) % 360.0
			local rightAngle = ( homingmissile.rot + maxTurnAngle ) % 360.0

			if(perpCosineAngle>0) then
				homingmissile.rot = rightAngle
			else
				homingmissile.rot = leftAngle
			end
		else
			homingmissile.rot = rot_tar
		end

		-- Work out new direction
		local rotrad	=	math.rad(homingmissile.rot)
		
		local vx = math.cos(rotrad) 
		local vy = math.sin(rotrad)
		
		homingmissile.vDir:SetXyzw(vx * dt * HomingMissile.kHMISSILE_SPEED,vy * dt * HomingMissile.kHMISSILE_SPEED,0,0)
		
		Vector4.Add(homingmissile.vPos, homingmissile.vDir, homingmissile.vPos)

		
	end

	function HomingMissile.GetCollisionLine(homingmissile,dt)
		--Logger.lprint("HomingMissile:GetCollisionLine()") 
		
		local future = Vector4.Create( homingmissile.vDir)
		Vector4.Multiply(future,future,4) 
		Vector4.Add(future,future,homingmissile.vPos)
		
		homingmissile.colline:SetPoints(homingmissile.vPos,future)
--		homingmissile.colline:CalculateCentreAndRadius() 

		return homingmissile.colline 
		
	end
	
	function HomingMissile.Render(homingmissile,renderHelper)
		--	Logger.lprint("HomingMissile:Render()") 

		RenderHelper.DrawTexture(renderHelper,homingmissile.animation:GetRenderTexture(),homingmissile.vPos:X(), homingmissile.vPos:Y(),-homingmissile.rot - 90,0.5)
	--	HomingMissile.RenderDebug(homingmissile,renderHelper)

	end

	function HomingMissile.RenderDebug(homingmissile,renderHelper)
		local mpos 	= 	homingmissile:GetVectorPosition()
 		RenderHelper.SetColour(renderHelper, {0, 0, 255, 128} )
		RenderHelper.DrawCircle(renderHelper,mpos:X(), mpos:Y(),  10)
		RenderHelper.DrawText(renderHelper,homingmissile:toString(),mpos:X()+40, mpos:Y())
		
		HomingMissile.GetCollisionLine(homingmissile):Render(renderHelper)
		
		
	end

	function HomingMissile.toString(homingmissile)
		local str =	{}
		return string.format("HomingMissile: state %d rot %f time %f pos = %s  diffV = %s",
			homingmissile.nState,
			homingmissile.rot,
			homingmissile.time,homingmissile.vPos:toString(),Temp.v0:toString())
	end
	
	function HomingMissile.GetVectorPosition(homingmissile)
		return homingmissile.vPos
	end

	function HomingMissile.GetCoords(homingmissile)
		return homingmissile.vPos:X(),homingmissile.vPos:Y(),homingmissile.vPos:Z()
	end
	
	function HomingMissile.GetPosition(homingmissile)
		return homingmissile.vPos:X(),homingmissile.vPos:Y(),homingmissile.vPos:Z()
	end
	
	function HomingMissile.SetPosition(homingmissile, v)
		homingmissile.vPos = Vector4.Create(v)
	end
	
	function HomingMissile.MarkForDeath(homingmissile)
		homingmissile.bMarkedForDeath = true
		ExplosionManager.AddExplosion({x=homingmissile.vPos:X(),y=homingmissile.vPos:Y(),z=0})
	end
	
	function HomingMissile.IsMarkedForDeath(homingmissile)
		return homingmissile.bMarkedForDeath
	end
	
	function HomingMissile.ApplyDamage(homingmissile,damage)
		homingmissile:MarkForDeath()
	end
	
	function HomingMissile.GetDamage(homingmissile)
		return homingmissile.damage
	end
	
	function HomingMissile.Cleanup(homingmissile)
		--Logger.lprint("HomingMissile.Render()") 
		homingmissile.vPos		=	nil
		homingmissile.following	=	nil
		homingmissile.tMissile	=	nil
		homingmissile.vDir		=	nil
		homingmissile.colline	=	nil 
	end

end

File: source/library/HomingMissile.js
//Copyright 2012, John Wilson, Brighton Sussex UK. Licensed under the BSD License. See licence.txt



var	STATE_HMISSILE_TRACKING					=	1 ;
var	STATE_HMISSILE_FALLING					=	2 ;
var	STATE_HMISSILE_SLEEPING					=	3 ;


var	kHMISSILES_MAX							= 	8 ;
var	kHMISSILE_MAXTURNANGLE					=	90.0 ;
var	kHMISSILE_SPEED							=	400.0 ;
var	kHMISSILE_LIFESPAN						=	15.0 ;
var	kHMISSILE_ANGULAR_ACCELERATION			=	180.0 ;


var	ANIM_ENEMYMISSILE_FRAME1				=	"enemy_missile__1" ;
var	ANIM_ENEMYMISSILE_FRAME2				=	"enemy_missile__2" ;
var	ANIM_ENEMYMISSILE_FRAME3				=	"enemy_missile__3" ;

var	Tempv0									=	Vector4.Create() ;
var	Tempv1									=	Vector4.Create() ;
var	Tempv2									=	Vector4.Create() ;
var	Tempv3									=	Vector4.Create() ;

function rad(deg)
{
	return deg*Math.PI/180 ;
}
function deg(rad)
{
	return rad*180/Math.PI ;
}

function HomingMissile( owner, startpos,  rot,  follow,  damage, lifespan)
{

		this.owner					=	owner ;
		this.following				=	follow ;
		this.vPos					=	Vector4.Create(startpos) ; 
		this.vDir					=	Vector4.Create(0,0) ;
        this.rot					=	rot || 0.0 ;
        this.Lifespan				=	lifespan || kHMISSILE_LIFESPAN ;
        this.time					=	0 ;
		this.strength				=	255 ;
        this.nState					=	STATE_HMISSILE_TRACKING ;
        this.damage					=	damage || 0 ;
        this.bInvincible			=	false ;
		this.bMarkedForDeath		=	false ;
		this.isPlayer 				=	(follow.className != 'Player') ;
		this.colline 				=	Line2D.Create(Vector4.Create(0,0),Vector4.Create(0,0)) ;
        this.currentAngularDir		=	Vector4.Create();
        this.currentAngularSpeed	=	Vector4.Create() ;
        this.tMissile 				=	null ;
		this.className				=	HomingMissile.className ;

		this.vPos.Add(this.vPos,Vector4.Create(4*Math.random(),4*Math.random())) ;

		this.animation 				= new TextureAnim() ;
		this.animation.addFrame(ANIM_ENEMYMISSILE_FRAME1) ;
		this.animation.addFrame(ANIM_ENEMYMISSILE_FRAME2) ;
		this.animation.addFrame(ANIM_ENEMYMISSILE_FRAME3) ;
		this.animation.play(true, true) ;

}

HomingMissile.className					=	"HomingMissile" ;

HomingMissile.Create			=	function( owner, startpos,  rot,  follow,  damage, lifespan)
{
	return new HomingMissile( owner, startpos,  rot,  follow,  damage, lifespan) ;
}

HomingMissile.prototype	=
{

	constructor:			HomingMissile,
	
	Follow:					function(who)
							{
								this.following = who ;
							},
	
	Update:					function(dt)
							{
						 		this.UpdateLive(dt) ;
							},
	
	UpdateLive:				function(dt)
							{
	
								this.time =	this.time +	dt ;

								if (this.animation) 
								{
							 		this.animation.update(dt) ;
								}
		
		
		
								var	diff_vec	=	Tempv0  ;
								var	state		=	this.nState ;
				
								if(state	!=	STATE_HMISSILE_FALLING && this.time > this.Lifespan) 
								{
									this.nState	=	STATE_HMISSILE_FALLING ;
									state		=	this.nState ;
								}
			   		
								if(state	==	STATE_HMISSILE_FALLING || !this.following.IsAlive()) 
								{
									
									diff_vec.SetXyzw(0,this.vPos.Y(),0,0) ;
									if (this.time > 1.25*this.Lifespan) 
									{
										this.nState		=	STATE_HMISSILE_SLEEPING ;
										this.MarkForDeath() ;
										this.time		=	0 ;
										return ;
									}
								}
								else if(state	==	STATE_HMISSILE_TRACKING) 
								{
									Vector4.Subtract(diff_vec,this.following.GetVectorPosition(), this.vPos) ;

								} else if(state	==	STATE_HMISSILE_SLEEPING) 
								{
									if (this.time > 4) 
									{
										//this.nState	=	STATE_HMISSILE_TRACKING ;
										this.time	=	0 ;
									}
									return ;
								}

								diff_vec.Normalise2() ;

								// Work out current angle
								var rot_tar =	Math.atan2(diff_vec.Y(), diff_vec.X()) ;
								rot_tar 	=	deg(rot_tar) ; 
								rot_tar 	=	(rot_tar + 360.0 ) % 360.0 ;


								var heading_vec= Tempv1 ;  	
		
								heading_vec.SetXyzw(this.vDir.X(),this.vDir.Y(),0,0) ;
								heading_vec.Normalise2() ;

								var perpheading_vec= Tempv2 ;

								perpheading_vec.SetXyzw(-heading_vec.Y(), heading_vec.X(), 0, 0) ;


								var perpCosineAngle = Vector4.Dot2(diff_vec, perpheading_vec) ;
		
								var newAngularDirection	=	(perpCosineAngle>0) && 1.0 || -1.0 ;
		
		
								if ( newAngularDirection != this.currentAngularDir ) 
								{
									this.currentAngularDir		=	newAngularDirection ;
									this.currentAngularSpeed	=	0.0 ;
								}

								this.currentAngularSpeed = this.currentAngularSpeed +(dt * kHMISSILE_ANGULAR_ACCELERATION) ;

								if ( this.currentAngularSpeed > kHMISSILE_MAXTURNANGLE ) 
								{
									this.currentAngularSpeed = kHMISSILE_MAXTURNANGLE ;
		
								}
		
								var maxTurnAngle = this.currentAngularSpeed * dt ;

								if(Math.abs(rot_tar-this.rot)>maxTurnAngle) 
								{
		
									var leftAngle  = ( this.rot + 360.0 - maxTurnAngle ) % 360.0 ;
									var rightAngle = ( this.rot + maxTurnAngle ) % 360.0 ;

									if(perpCosineAngle>0) {
										this.rot = rightAngle ;
									}
									else
									{
										this.rot = leftAngle ;			
									}
								}
								else
								{
									this.rot = rot_tar ;
		
								}

								// Work out new direction
								var rotrad	=	rad(this.rot) ;
		
								var vx = Math.cos(rotrad) ;
								var vy = Math.sin(rotrad) ;
		
								this.vDir.SetXyzw(vx * dt * kHMISSILE_SPEED,vy * dt * kHMISSILE_SPEED,0,0) ;
		
								Vector4.Add(this.vPos, this.vDir, this.vPos) ;

		
							},

	GetCollisionLine:		function(dt)
							{
							 	var	future = Vector4.Create( this.vDir) ;
								Vector4.Multiply(future,future,4) ;
								Vector4.Add(future,future,this.vPos) ;
		
								this.colline.SetPoints(this.vPos,future) ;
						//		this.colline.CalculateCentreAndRadius() 

								return this.colline ;
							},
		

	Render:					function(renderHelper)
							{
								if(this.nState	!=	STATE_HMISSILE_SLEEPING) 
								{
									renderHelper.drawRawImage(ImageResource.getResource(this.animation.getRenderTexture()),this.vPos.X(), this.vPos.Y(),90+this.rot,0.65,0.65) ;
									//this.RenderDebug(renderHelper)
								}
							},
	
	RenderDebug:			function(renderHelper)
							{
								var mpos 	= 	this.GetVectorPosition() ;
								renderHelper.drawCircle(mpos.X(), mpos.Y(),  4,[255,255,0]) ;
								
								renderHelper.drawText(this.toString(),mpos.X()+40, mpos.Y()) ;
								this.GetCollisionLine().Render(renderHelper) ;
	
							},
		
	toString:				function()
							{
								return "HomingMissile" ;
							},

	GetVectorPosition:		function()
							{
								return this.vPos ;
							},

	GetPosition:			function()
							{
								return {x: this.vPos.X(), y: this.vPos.Y(), z: this.vPos.Z()} ;
							},

	GetCoords:				function()
							{
								return this.GetPosition() ; 
							},
	
	
	SetPosition:			function(v)
							{
								this.vPos = Vector4.Create(v) ;
							},
	
	MarkForDeath:			function()
							{
								this.bMarkedForDeath = true ;
								var animationinfo = {texture:"bullethits",frameshor:4,framesvert:2,x:this.vPos.X(),y:this.vPos.Y(),loop:false,bounce:false,scale:4} ;
								ExplosionManager.AddExplosion(animationinfo,7) ;
							},
	
	IsMarkedForDeath:		function()
							{
								return this.bMarkedForDeath ;
	
							},
	
	ApplyDamage:			function(damage)
							{
								this.MarkForDeath()
	
							},
	
	GetDamage:				function()
							{
								return this.damage ;
							},
	

	IsAlive:				function()
							{
								return !this.IsMarkedForDeath() ;
							},
							
	Cleanup:				function()
							{
								this.vPos		=	null ;
								this.following	=	null ;
								this.tMissile	=	null ;
								this.vDir		=	null ;
								this.colline	=	null  ;
							},

}

MissileManager				=	{} 	;

MissileManager.follow		=	null ;
var hmissile				=	null ;

	
MissileManager.Init			=	function(follow)
								{
									hmissile	=	null ;
								}

MissileManager.CanFire		=	function()
								{
									return !hmissile ;
								}
								
MissileManager.AddMissile	=	function(owner, startpos,  rot,  follow,  damage, lifespan)
								{
									if (!hmissile) {
										hmissile			=	new HomingMissile(owner, startpos,  rot,  follow,  damage || 1, lifespan || 15) ;
									}
										
								}
								
MissileManager.Update		=	function(dt)
								{
									if (hmissile) {
										hmissile.Update(dt) ;	
										if (!hmissile.IsAlive()) 
										{
											hmissile	=	null ;
										}
									}

								}

MissileManager.Render		=	function(rHelper)
								{
									if (hmissile) {
										hmissile.Render(rHelper) ;	
									}
								}
								
